File: mobile/src/tools/ocr.ts
import { BaseTool } from './registry';
import { getLatestImage } from '../core/imageStore';
import { analyzeImage } from '../core/multimodalClient';

export class OcrTool extends BaseTool {
  name = 'ocr';
  description = '识别用户上传的图片中的文字和题目。当用户拍照或从相册选择图片后，调用此工具分析图片。对于几何图等复杂图形会直接解答。';
  parameters = {
    type: 'object',
    properties: {},
    required: [],
  };

  async execute(_kwargs: Record<string, any>): Promise<string> {
    const imageBase64 = getLatestImage();
    if (!imageBase64) {
      return '当前没有可用的图片。请先让用户使用相机拍照或从相册选择图片。';
    }

    const result = await analyzeImage(imageBase64);

    if (result.isTakeover) {
      return `__TAKEOVER__${result.text}`;
    }
    return `用户上传的图片中的题目：${result.text}`;
  }
}

File: backend/app/tools/ocr.py
from __future__ import annotations

from app.tools.registry import BaseTool


class OcrTool(BaseTool):
    name: str = "ocr"
    description: str = "识别图片中的文字和数学公式。支持手写和印刷体。"
    parameters: dict = {
        "type": "object",
        "properties": {
            "image_url": {
                "type": "string",
                "description": "图片的 URL 或 base64 data URI",
            },
            "language": {
                "type": "string",
                "description": "语言，默认 'chi_sim+eng'",
                "default": "chi_sim+eng",
            },
        },
        "required": ["image_url"],
    }

    async def execute(self, image_url: str, language: str = "chi_sim+eng") -> str:
        # 后续集成 PaddleOCR 或调用多模态模型
        # 当前返回占位信息
        return (
            "OCR 功能正在集成中。当前支持：\n"
            "1. 通过多模态模型直接识别图片内容\n"
            "2. 后续将集成 PaddleOCR 做本地识别\n\n"
            f"图片已接收 (长度: {len(image_url)} 字符)，请使用 AI 的多模态能力分析图片内容。"
        )
